services/gateway/shared/registerHttpApi.js

```diff
@@ -64,137 +94,143 @@ const configureApi = (apiNames, apiPrefix, registeredModuleNames) => {
+			res.end(data.data);
+			return res;
+		}
+
+		if (data.data && data.status) {
+			if (data.status === 'SERVICE_UNAVAILABLE') ctx.meta.$responseHeaders = { 'Retry-After': 30 };
+			if (data.status === 'INVALID_PARAMS') data.status = 'BAD_REQUEST';
+
+			ctx.meta.$statusCode = StatusCodes[data.status] || data.status;
+
+			// 'ACCEPTED' is a successful HTTP code
+			if (data.status !== 'ACCEPTED') {
+				let message = `The request ended up with ${data.status} error`;
+
+				if (typeof data.data === 'object' && typeof data.data.error === 'string') {
+					message = data.data.error;
+				}
 
-		async onBeforeCall(ctx, route, req, res) {
-			const sendResponse = (code, message) => {
-				res.setHeader('Content-Type', 'application/json');
-				res.writeHead(code || 400);
-				res.end(JSON.stringify({
+				return {
 					error: true,
 					message,
-				}));
-			};
-
-			const routeAlias = `${req.method.toUpperCase()} ${req.$alias.path}`;
-			const paramReport = validate(req.$params, methodPaths[routeAlias]);
-
-			if (paramReport.missing.length > 0) {
-				sendResponse(INVALID_REQUEST[0], `Missing parameter(s): ${paramReport.missing.join('; ')}`);
-				throw new ValidationException('Request param validation error.');
-			}
-
-			const unknownList = Object.keys(paramReport.unknown);
-			if (unknownList.length > 0) {
-				sendResponse(INVALID_REQUEST[0], `Unknown input parameter(s): ${unknownList.join('; ')}`);
-				throw new ValidationException('Request param validation error.');
-			}
-
-			if (paramReport.required.length) {
-				sendResponse(INVALID_REQUEST[0], `Require one of the following parameter combination(s): ${paramReport.required.join('; ')}`);
-				throw new ValidationException('Request param validation error.');
-			}
-
-			const invalidList = paramReport.invalid;
-			if (invalidList.length > 0) {
-				sendResponse(INVALID_REQUEST[0], `Invalid input: ${invalidList.map(o => o.message).join('; ')}`);
-				throw new ValidationException('Request param validation error.');
+				};
 			}
+		}
+		const apiPath = `${req.method.toUpperCase()} ${req.$alias.path}`;
+		return transformResponse(methodPaths[apiPath], data);
+	},
+});
 
-			const params = transformRequest(methodPaths[routeAlias], dropEmptyProps(paramReport.valid));
-			req.$params = params;
-		},
+const registerApi = (apiNames, config, registeredModuleNames) => {
+	const allAPIs = getAllAPIs(apiNames, registeredModuleNames);
 
-		async onAfterCall(ctx, route, req, res, data) {
-			if (gatewayConfig.api.enableHttpCacheControl) {
-				// Set 'Cache-Control' to assist response caching
-				ctx.meta.$responseHeaders = { 'Cache-Control': gatewayConfig.api.httpCacheControlDirectives };
-			}
+	const falseEtagAPIs = Object.fromEntries(Object.entries(allAPIs)
+		.filter(([, value]) => value.etag !== undefined && !value.etag));
+	const strongEtagAPIs = Object.fromEntries(Object.entries(allAPIs)
+		.filter(([, value]) => value.etag === undefined || value.etag === 'strong'));
```

Comment:
`etag` can contain one of the five possible values as specified [here](https://moleculer.services/docs/0.14/moleculer-web.html#ETag). So, we'll need to handle all the five scenarios. In case of a custom function, we'll need to execute it in L115 inside of `getAPIConfig`.